topdown/providers/aws.go

```diff
@@ -0,0 +1,210 @@
+// Copyright 2022 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package providers
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"fmt"
+	"net/url"
+	"sort"
+	"strings"
+	"time"
+
+	ast "github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/topdown/builtins"
+)
+
+func stringFromTerm(t *ast.Term) string {
+	out := ""
+	if v, ok := t.Value.(ast.String); ok {
+		return string(v)
+	}
+	return out
+}
+
+// --------------------------------------------------------
+// AWS
+
+// AWS-specific helper values for use in topdown/builtins.
+// Directly adapted from plugins/rest/aws.go where possible.
+
+// const (
+// 	// ref. https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-envvars.html
+// 	accessKeyEnvVar               = "AWS_ACCESS_KEY_ID"
+// 	secretKeyEnvVar               = "AWS_SECRET_ACCESS_KEY"
+// 	securityTokenEnvVar           = "AWS_SECURITY_TOKEN"
+// 	sessionTokenEnvVar            = "AWS_SESSION_TOKEN"
+// 	awsRegionEnvVar               = "AWS_REGION"
+// 	awsRoleArnEnvVar              = "AWS_ROLE_ARN"
+// 	awsWebIdentityTokenFileEnvVar = "AWS_WEB_IDENTITY_TOKEN_FILE"
+// 	awsCredentialsFileEnvVar      = "AWS_SHARED_CREDENTIALS_FILE"
+// 	awsProfileEnvVar              = "AWS_PROFILE"
+// )
+
+type awsCredentials struct {
+	AccessKey    string
+	SecretKey    string
+	RegionName   string
+	SessionToken string
+}
+
+// Headers that may be mutated before reaching an aws service (eg by a proxy) should be added here to omit them from
+// the sigv4 canonical request
+// ref. https://github.com/aws/aws-sdk-go/blob/master/aws/signer/v4/v4.go#L92
+var awsSigv4IgnoredHeaders = map[string]bool{
+	"authorization":   true,
+	"user-agent":      true,
+	"x-amzn-trace-id": true,
+}
```

Comment:
[nit] It's quite common to use `struct{}` here instead,
```suggestion
var awsSigv4IgnoredHeaders = map[string]struct{}{
	"authorization":   {},
	"user-agent":      {},
	"x-amzn-trace-id": {},
}
```